### PR TITLE
Iterable ObjectStorage interface for use in Repository struct

### PR DESCRIPTION
**Pull request by JoshuaSjoding**
The current implementation of the Repository struct includes a [pointer to core.RAWObjectStorage](https://github.com/src-d/go-git/blob/e9164d853bed12ed9f6787a608db32504c894a64/repository.go#L22). This seems inappropriate because it prevents the development of alternative implementations of [ObjectStorage](https://github.com/src-d/go-git/blob/e9164d853bed12ed9f6787a608db32504c894a64/core/object.go#L20). I presume that the direct reference was included because it provided convenient and fast access to the commit objects.

This pull request proposes the addition of one function to the ObjectStorage interface: [IterType](https://github.com/scjalliance/go-git/blob/4172d931089d9851614217df7c8f57693587399e/core/object.go#L24). This function returns an iterator of objects matching the given object type. The Repository is modified to make use of this new function, and RAWObjectStorage is modified to provide it.

The return value of IterType is a new interface called [ObjectIter](https://github.com/scjalliance/go-git/blob/4172d931089d9851614217df7c8f57693587399e/core/object.go#L57). Two implementations of ObjectIter are provided:

* ObjectLookupIter, which iterates over a slice of object hashes and looks up the given objects as needed. This is used to iterate through parent commits, for example.
* ObjectSliceIter, which iterates over a slice of objects.

Additional implementations could easily be added.

Both CommitIter and TreeIter are modified to make use of the new ObjectIter interface, as is Commit.Parents(). Note that this does away with the previous channel-based iterators.

The proposed [implementation of RAWObjectStorage.IterType](https://github.com/scjalliance/go-git/blob/4172d931089d9851614217df7c8f57693587399e/core/object.go#L192) simply flattens the map of the requested type into a slice on the fly. This is probably fine for small to medium repositories, but *might* cause trouble for large repositories or if it's called frequently. Further optimization should be done as warranted.

Note that this pull request **modifies several public types**. As such it should probably be considered breaking for v2; it may be more appropriate to pull this instead into a v3 branch. I expect that other breaking changes will be needed to meet the needs of #19, and I suspect that all such work should be carried out in a branch until it's ready to be rolled out formally as v3. Feedback on this is most welcome.

### CODE REVIEW

File: core/object.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ type ObjectStorage interface {
 	New() Object
 	Set(Object) Hash
 	Get(Hash) (Object, bool)
+	IterType(ObjectType) ObjectIter
```
Comment (mcuadros) [inline core/object.go]: I prefer call this method just `Iter`, is less verbose and clear enough

File: core/object.go [hunk 2]
```diff
@@ -52,6 +53,82 @@ func (t ObjectType) Bytes() []byte {
...
 }
 
+// ObjectIter is a generic closable interface for iterating over objects.
+type ObjectIter interface {
+	Next() (Object, error)
+	Close()
+}
+
+// ObjectLookupIter yields a series of objects by retrieving each one from
+// object storage.
```
Comment (mcuadros) [inline core/object.go]: please explain that is an implementation of `ObjectIter`

File: core/object.go [hunk 3]
```diff
@@ -52,6 +53,82 @@ func (t ObjectType) Bytes() []byte {
...
+// the end it will return io.EOF as an error. If the object is retreieved
+// successfully error will be nil.
+func (iter *ObjectLookupIter) Next() (Object, error) {
+	if iter.pos >= len(iter.series) {
+		return nil, io.EOF
+	}
+	hash := iter.series[iter.pos]
+	obj, ok := iter.storage.Get(hash)
+	if !ok {
+		// FIXME: Consider making ObjectStorage.Get return an actual error that we
```
Comment (mcuadros) [inline core/object.go]: Make sense, you can elaborate another PR with this change.

File: commit_test.go [hunk 4]
```diff
@@ -40,11 +40,14 @@ func (s *SuiteCommit) SetUpSuite(c *C) {
 	}
 }
 
+// FIXME: Test the new CommitIter
```
Comment (mcuadros) [inline commit_test.go]: what about this test?

**Comment by mcuadros** [comment]
@JoshuaSjoding great job, we will merge this as soon as the minor things I commented get fixed. You can keep the PR to master, because we are using gopkg.in, is not a problem having an unreleased master.

Thanks!

File: core/object.go (hunk 3 above)
Comment (JoshuaSjoding) [inline core/object.go]: On the one hand I like having `ObjectStorage.Get()` use the same interface as a standard map, but on the other hand we might want to report more detailed errors from an underlying file system.

Either way, I agree that it should be addressed in a separate PR. Having a functional `FileObjectStorage` implementation will also be informative.

Comment (JoshuaSjoding) [inline core/object.go]: In the meantime I made `ObjectLookupIter.Next()` return an `ObjectNotFoundErr` when the underlying storage provider can't return the requested object. This way it can at least be distinguished from a regular `io.EOF`.

**Comment by JoshuaSjoding** [comment]
@mcuadros I think that I've addressed all of your comments with the latest commit.

`CommitIter` has some decent test coverage now. `Tree` probably needs some work and should use a `FileIter` to be consistent, but I'll handle that with a separate issue and PR if it's wanted.

I also merged in the latest work from `src-d:master` for convenience. Let me know if you'd prefer that I not do that in the future.

File: core/object.go (hunk 3 above)
Comment (mcuadros) [inline core/object.go]: Make sense having a more verbose output of  `ObjectStorage.Get()` because now a simple call to it can fail in many different way if is base for example in filesystem.

**Comment by JoshuaSjoding** [comment]
@mcuadros I went ahead and updated the `ObjectStorage` functions `New`, `Get` and `Set` so that they can return errors now.
